src/backend/metalll/src/command.rs

```diff
@@ -0,0 +1,613 @@
+    }
+}
+
+pub struct RenderPassInlineEncoder<'cb, 'rp, 'fb, 'enc: 'cb> {
+    command_buffer: &'cb mut command::Encoder<'enc, CommandBuffer>,
+    render_pass: &'rp native::RenderPass,
+    framebuffer: &'fb native::FrameBuffer,
+    render_encoder: MTLRenderCommandEncoder,
+}
+
+impl<'cb, 'rp, 'fb, 'enc> Drop for RenderPassInlineEncoder<'cb, 'rp, 'fb, 'enc> {
+    fn drop(&mut self) {
+        unsafe {
+            self.render_encoder.end_encoding();
+            self.render_encoder.release();
+        }
+    }
+}
+
+pub struct RenderPassSecondaryEncoder<'cb, 'rp, 'fb, 'enc: 'cb> {
+    command_buffer: &'cb mut command::Encoder<'enc, CommandBuffer>,
+    render_pass: &'rp native::RenderPass,
+    framebuffer: &'fb native::FrameBuffer,
+}
+
+impl<'cb, 'rp, 'fb, 'enc> command::RenderPassEncoder<'cb, 'rp, 'fb, 'enc, CommandBuffer, Resources> for RenderPassInlineEncoder<'cb, 'rp, 'fb, 'enc>
+{
+    type SecondaryEncoder = RenderPassSecondaryEncoder<'cb, 'rp, 'fb, 'enc>;
+    type InlineEncoder = RenderPassInlineEncoder<'cb, 'rp, 'fb, 'enc>;
+
+    fn begin(command_buffer: &'cb mut Encoder<'enc, CommandBuffer>,
+             render_pass: &'rp native::RenderPass,
+             framebuffer: &'fb native::FrameBuffer,
+             render_area: target::Rect,
+             clear_values: &[ClearValue]) -> Self
+    {
+        unsafe {
+            // FIXME: subpasses
+            
+            let pass_descriptor = framebuffer.0;
+            // TODO: we may want to copy here because we will modify the Framebuffer (by setting
+            // clear colors)
+            // TODO: validate number of clear colors
+            for (i, value) in clear_values.iter().enumerate() {
+                let color_desc = pass_descriptor.color_attachments().object_at(i);
+                let mtl_color = match *value {
+                    ClearValue::Color(ClearColor::Float(values)) => MTLClearColor::new(
+                        values[0] as f64,
+                        values[1] as f64,
+                        values[2] as f64,
+                        values[3] as f64,
+                    ),
+                    _ => unimplemented!(),
+                };
+                color_desc.set_clear_color(mtl_color);
+            }
+
+            let render_encoder = command_buffer.command_buffer.new_render_command_encoder(pass_descriptor);
+            defer_on_unwind! { render_encoder.release() };
+
+            // Apply previously bound values for this command buffer
+            if let Some(viewport) = command_buffer.viewport {
+                render_encoder.set_viewport(viewport);
+            }
+            if let Some(scissors) = command_buffer.scissors {
+                render_encoder.set_scissor_rect(scissors);
+            }
+            if let Some(pipeline_state) = command_buffer.pipeline_state {
+                render_encoder.set_render_pipeline_state(pipeline_state);
+            } else {
+                panic!("missing bound pipeline state object");
```

Comment:
Is this a requirement for Metal Encoders to have always an bound PSO?
I've seen various other constraints in some backends (e.g vertex bind slots need to defined etc.).
We should document them somewhere but that's out of scope.